middleware/recovery.go

```diff
@@ -0,0 +1,129 @@
+package middleware
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"runtime/debug"
+	"sync"
+
+	"github.com/danielgtaylor/huma"
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+)
+
+// MaxLogBodyBytes logs at most this many bytes of any request body during a
+// panic when using the recovery middleware. Defaults to 10KiB.
+var MaxLogBodyBytes int64 = 10 * 1024
+
+// bufferedReadCloser will read and buffer up to max bytes into buf. Additional
+// reads bypass the buffer.
+type bufferedReadCloser struct {
+	reader io.ReadCloser
+	buf    *bytes.Buffer
+	max    int64
+}
+
+// newBufferedReadCloser returns a new BufferedReadCloser that wraps reader
+// and reads up to max bytes into the buffer.
+func newBufferedReadCloser(reader io.ReadCloser, buffer *bytes.Buffer, max int64) *bufferedReadCloser {
+	return &bufferedReadCloser{
+		reader: reader,
+		buf:    buffer,
+		max:    max,
+	}
+}
+
+// Read data into p. Returns number of bytes read and an error, if any.
+func (r *bufferedReadCloser) Read(p []byte) (n int, err error) {
+	// Read from the underlying reader like normal.
+	n, err = r.reader.Read(p)
+
+	// If buffer isn't full, add to it.
+	length := int64(r.buf.Len())
+	if length < r.max {
+		if length+int64(n) < r.max {
+			r.buf.Write(p[:n])
+		} else {
+			r.buf.Write(p[:int64(n)-(r.max-length)])
+		}
+	}
+
+	return
+}
+
+// Close the underlying reader.
+func (r *bufferedReadCloser) Close() error {
+	return r.reader.Close()
+}
+
+// PanicFunc defines a function to run after a panic, which allows you to set
+// up custom logging, metrics, etc.
+type PanicFunc func(ctx context.Context, err error, request string)
+
+// Recovery prints stack traces on panic when used with the logging middleware.
+func Recovery(onPanic PanicFunc) func(http.Handler) http.Handler {
+	bufPool := sync.Pool{
+		New: func() interface{} {
+			return new(bytes.Buffer)
+		},
+	}
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			var buf *bytes.Buffer
+
+			// Reset the body so other middleware/handlers can use it.
+			if r.Body != nil {
+				// Get a buffer that the body will be read into.
+				buf = bufPool.Get().(*bytes.Buffer)
+				defer bufPool.Put(buf)
+
+				r.Body = newBufferedReadCloser(r.Body, buf, MaxLogBodyBytes)
+			}
+
+			// Recovering comes *after* the above so the buffer is not returned to
+			// the pool until after we print out its contents.
+			defer func() {
```

Comment:
Anonymous goroutines are hard to test independently.

Reply:
This is just the standard way to do recovery and it is tested in `recovery_test.go`.

Reply:
I accept that a `defer func()` is the standard way to do recovery. I'm talking more broadly about the fact that nested anonymous functions can't be tested in isolation of their surrounding code. That to me is a code smell, especially when the fix is as simple as "declare the function elsewhere in the code and reference it by name." I accept that sometimes this is going to happen in places where we really do intend these components really _shouldn't_ be decomposed, but I'd prefer it if in our codebase those are the exception and called out with comments, rather than the rule.